src/com/xilinx/rapidwright/interchange/MacroParamRule.java

```diff
@@ -0,0 +1,68 @@
+package com.xilinx.rapidwright.interchange;
+
+import com.xilinx.rapidwright.design.Unisim;
+
+public class MacroParamRule {
+    public String primParam;
+    public String instName;
+    public String instParam;
+    public int[] bitSlice;
+    public MacroParamTableEntry[] tableLookup;
+
+    public MacroParamRule(String primParam, String instName, String instParam) {
+        this.primParam = primParam;
+        this.instName = instName;
+        this.instParam = instParam;
+        this.bitSlice = null;
+        this.tableLookup = null;
+    }
+
+    public static MacroParamRule copy(String primParam, String instName, String instParam) {
+        return new MacroParamRule(primParam, instName, instParam);
+    }
+
+    public static MacroParamRule bit_range(String primParam, String instName, String instParam, int start, int count) {
+        MacroParamRule result = new MacroParamRule(primParam, instName, instParam);
+        result.bitSlice = new int[count];
+        for (int i = 0; i < count; i++)
+            result.bitSlice[i] = start + i;
+        return result;
+    }
+
+    public static MacroParamRule table(String primParam, String instName, String instParam, MacroParamTableEntry[] table) {
+        MacroParamRule result = new MacroParamRule(primParam, instName, instParam);
+        result.tableLookup = table;
+        return result;
+    }
+
+    public static MacroParamRule[] getRules(Unisim prim) {
+        switch (prim) {
```

Comment:
Thanks @gatecat for the work.  I am wondering if we could avoid hardcoding the data.  I am looking at extracting it from the Macro EDIFLibrary netlist which seems to have mostly all the information (easy way to dump to a file):
```
Design.getMacroPrimitives(Series.Series7).exportEDIF(new FileWriter("Series7_macros.edf"));
```
Based on the examples, the only thing I can't seem to derive is the `start` of `bitRange`.  Perhaps you could describe how you inferred those values for your examples and we could device a scheme to derive or augment existing RapidWright data to supply them.

Reply:
As an example of the way to infer these, consider the following Tcl:

```tcl
link_design -part xc7a35tftg256-1
create_cell -reference RAM128X1D ram_i
set_property INIT 128'hFFEEDDCCBBAA99887766554433221100 [get_cells ram_i]
puts "LOW: [get_property INIT [get_cells ram_i/DP.LOW]]"
puts "HIGH: [get_property INIT [get_cells ram_i/DP.HIGH]]"
```

prints
```
LOW: 64'h7766554433221100
HIGH: 64'hFFEEDDCCBBAA9988
```

it would definitely be even better to have this info available directly in RapidWright! I'd also be curious if the placement rules for macros are something that could be included as part of the RapidWright in the longer term (I think Vivado calls these "shapes" based on the error messages when you get it wrong.)